loopdb/sqlite.go

```diff
@@ -191,6 +206,82 @@ func (db *BaseDB) ExecTx(ctx context.Context, txOptions TxOptions,
 	return nil
 }
 
+// FixFaultyTimestamps fixes faulty timestamps in the database, caused
+// by using milliseconds instead of seconds as the publication deadline.
+func (b *BaseDB) FixFaultyTimestamps(ctx context.Context,
+	parseTimeFunc func(string) time.Time) error {
+
+	// Manually fetch all the loop out swaps.
+	rows, err := b.DB.QueryContext(
+		ctx, "SELECT swap_hash, publication_deadline FROM loopout_swaps",
+	)
+	if err != nil {
+		return err
+	}
+
+	// Parse the rows into a struct. We need to do this manually because
+	// the sqlite driver will fail on faulty timestamps.
+	type LoopOutRow struct {
+		Hash                []byte `json:"swap_hash"`
+		PublicationDeadline string `json:"publication_deadline"`
+	}
+
+	var loopOutSwaps []LoopOutRow
+
+	for rows.Next() {
+		var swap LoopOutRow
+		err := rows.Scan(
+			&swap.Hash, &swap.PublicationDeadline,
+		)
+		if err != nil {
+			return err
+		}
+
+		loopOutSwaps = append(loopOutSwaps, swap)
+	}
+
+	tx, err := b.BeginTx(ctx, &SqliteTxOptions{})
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback() //nolint: errcheck
+
+	for _, swap := range loopOutSwaps {
+		faultyTime := parseTimeFunc(swap.PublicationDeadline)
+
+		// Skip if the time is not faulty.
+		if !isMilisecondsTime(faultyTime.Unix()) {
+			continue
+		}
+
+		// Update the faulty time to a valid time.
+		secs := faultyTime.Unix() / 1000
+		nsecs := (faultyTime.Unix() % 1000) * 1e6
```

Comment:
Since we currently ignore anything less than a second in the parse functions we'll always end up with zero nanosecs.